### DISCUSSION TITLE
Ctrl-C doesn't work with non-latin input source

### DISCUSSION THREAD

**Discussion by kolontsov** [ghostty-org/ghostty#2757] [Q&A] [answered]
macOS 15.1, Karabiner-Elements 15.3, ghostty version 63bf16ff

Hi, my Ctrl-C, Ctrl-D (and others) don't work when I use non-US input source. Here is example what I see in `cat`: `^[[1089;5u` for ^C,  `^[[1074;5u` for ^D - so it's unicode code of Cyrillic letters. I checked standard macos terminal and ctrl-based combinations work as expected regardless the input source.

**Comment by mitchellh** [comment]
Can you please tell me what keyboard layout you’re using and what physical keys on a standard US layout keyboard you’re pressing?

**Reply by kolontsov** [reply]
Sure, 
```
$  defaults read /Library/Preferences/com.apple.HIToolbox.plist  AppleEnabledInputSources
(
        {
        InputSourceKind = "Keyboard Layout";
        "KeyboardLayout ID" = 252;
        "KeyboardLayout Name" = ABC;
    },
        {
        InputSourceKind = "Keyboard Layout";
        "KeyboardLayout ID" = 19456;
        "KeyboardLayout Name" = Russian;
    }
)
```

Latin "C" and russian "С" (yes, they look totally the same) are on the same physical key. So I press Control and 3rd key in "zxcvbn" row.

**Reply by kolontsov** [reply]
Just in case, that's how pressing Ctrl+C looks in Karabiner-EventViewer with russian input source being active (so key code is correct):
```
[
  {
    "type": "down",
    "name": {"key_code":"left_control"},
    "usagePage": "7 (0x0007)",
    "usage": "224 (0x00e0)",
    "misc": "flags left_control"
  },
  {
    "type": "down",
    "name": {"key_code":"c"},
    "usagePage": "7 (0x0007)",
    "usage": "6 (0x0006)",
    "misc": "flags left_control"
  },
  {
    "type": "up",
    "name": {"key_code":"c"},
    "usagePage": "7 (0x0007)",
    "usage": "6 (0x0006)",
    "misc": "flags left_control"
  },
  {
    "type": "up",
    "name": {"key_code":"left_control"},
    "usagePage": "7 (0x0007)",
    "usage": "224 (0x00e0)",
    "misc": ""
  }
]
```

**Reply by kolontsov** [reply]
I guess my issue can be resolved by using `ctrl+physical:c` keybind, but how to send "proper" Ctrl-C? `text:\x3` doesn't seem to work.

**Reply by mitchellh** [reply]
> I guess my issue can be resolved by using ctrl+physical:c keybind, but how to send "proper" Ctrl-C? text:\x3 doesn't seem to work.

That should work. I'm not sure if `\x3` is the right syntax but something like that should work. 

I still have it on my TODO to try to reproduce this and see if its an issue or not.

**Reply by kolontsov** [reply]
> That should work. I'm not sure if \x3 is the right syntax but something like that should work.

Thank you! Right syntax is `\x03` (I feel silly now), so `keybind = ctrl+physical:c=text:\x03` is the perfect workaround.

**Reply by markpeek** [reply]
> I still have it on my TODO to try to reproduce this and see if its an issue or not.

@mitchellh I see this is an older discussion that got updated but is the same as #3513 and #3584. The simple repro on MacOS is:

1. With a US keyboard, add a Russian input source from System Settings > Keyboard > Input Sources > Edit
2. Switch to Russian in the Input menu
3. Hit ctrl-c in ghostty

I had these [comments](https://github.com/ghostty-org/ghostty/issues/3513#issuecomment-2564559867) on the cause.

**Reply by mitchellh** [reply, marked as answer]
This was fixed: #4518

**Comment by esn89** [comment]
I have a similar issue:

ctrl+c doesn't seem to work, but I have a regular english (US) input.  The only caveat being that I am using Karabiner's modification where I turn my Caps Lock into:

Escape when tapped.
But, it becomes "ctrl" when I use it with a combination.  Such as Ctrl-C.

However, my Ctrl-C doesn't seem to work, nor Ctrl-R, for reverse history search.

**Comment by esn89** [comment]
The Karabiner Events output is:

```json
[
  {
    "type": "down",
    "name": {"key_code":"left_control"},
    "usagePage": "7 (0x0007)",
    "usage": "224 (0x00e0)",
    "misc": "flags left_control"
  },
  {
    "type": "up",
    "name": {"key_code":"left_control"},
    "usagePage": "7 (0x0007)",
    "usage": "224 (0x00e0)",
    "misc": ""
  },
  {
    "type": "down",
    "name": {"key_code":"left_command"},
    "usagePage": "7 (0x0007)",
    "usage": "227 (0x00e3)",
    "misc": "flags left_command"
  },
  {
    "type": "down",
    "name": {"key_code":"c"},
    "usagePage": "7 (0x0007)",
    "usage": "6 (0x0006)",
    "misc": "flags left_command"
  },
  {
    "type": "up",
    "name": {"key_code":"c"},
    "usagePage": "7 (0x0007)",
    "usage": "6 (0x0006)",
    "misc": "flags left_command"
  },
  {
    "type": "up",
    "name": {"key_code":"left_command"},
    "usagePage": "7 (0x0007)",
    "usage": "227 (0x00e3)",
    "misc": ""
  }
]
```

Is there any way around this?

**Reply by Tylermarques** [reply]
I was using Karabiner but I uninstalled it and am still experiencing the same issue. ctrl+c works in Terminal.app and Alacritty. The other control sequences work fine (ctrl+r, ctrl+t, ctrl+z) in all three. The ctrl+c sequence is the only one I'd remapped using karabiner which is suspicious.

Using `showkey` I don't actually see any output in Ghostty when pressing ctrl+c. I would expect ctrl+c to exit showkey, as it does in Alacritty and Terminal.app. Turning on "Secure Input" doesn't have an effect on the behaviour. [ I've included a recording of Ghostty and Alacritty side by side.](https://youtu.be/o2SuuBF6w0w)

I'm on version 1.0.1 of Ghostty, commit [3f7c3afa](https://github.com/ghostty-org/ghostty/commits/3f7c3afa)

Let me know if there is other information I can provide or other debugging steps you can think of. Really appreciate all the work the Ghostty team does. Great emulator :) Thank you!

**Reply by Tylermarques** [reply]
Alright I've discovered the issue. Ghostty seems to respect the settings found in "System Preferences > Keyboard > Keyboard Shortcuts > App Shortcuts." where as Alacritty doesn't. 

I do not know how it got changed, but ctrl+c was mapped there to "Copy". Removing that fixes the issue completely.
